server/src/main/java/io/druid/client/CachingClusteredClient.java

```diff
@@ -266,6 +266,17 @@ public CachingClusteredClient(
 
       final Set<ServerToSegment> segments = computeSegmentsToQuery(timeline);
       @Nullable final byte[] queryCacheKey = computeQueryCacheKey();
+
+      if (responseContext.containsKey("resultLevelCacheEnabled")) {
+        @Nullable final String prevResultCacheKeyFromEtag = (String) responseContext.get("prevResultSetIdentifier");
+        @Nullable final String newResultCacheKeyFromEtag = computeCurrentEtag(segments, queryCacheKey);
+        responseContext.remove("resultLevelCacheEnabled");
+        responseContext.put("currentResultSetIdentifier", newResultCacheKeyFromEtag);
+        if (newResultCacheKeyFromEtag != null && newResultCacheKeyFromEtag.equals(prevResultCacheKeyFromEtag)) {
+          return Sequences.empty();
+        }
+      }
```

Comment:
this kind of short circuiting is already done by checking `QueryResource.HEADER_IF_NONE_MATCH` in the query context can we reuse same and not add this block ?

Reply:
Fixed as per suggestion.